### ISSUE TITLE
How to get message_state from deliver_sm in hook?

### ISSUE THREAD

**Issue by Yolley** [komuw/naz#158]
## Which version of python are you using?
3.7

## What operating system and version of operating system are you using?
Centos 7

## What version of naz are you using? `naz-cli --version`
naz v0.6.6

## What did you do? (be as detailed as you can)
Made simple hook with logging
```python
class Hook(BaseHook):
    async def request(self, smpp_command, log_id, hook_metadata):
        REQUESTS_COUNTER.labels(command=smpp_command, env=config.ENVIRONMENT).inc()

    async def response(self, smpp_command, log_id, hook_metadata, smsc_response):
        logger.info(f'Got smpp_command {smpp_command} and log_id {log_id}, smsc_response {smsc_response}')
        RESPONSES_COUNTER.labels(command=smpp_command, env=config.ENVIRONMENT).inc()
```
Sent message and received deliver_sm for this message with message_state (http://docs.nimta.com/SMPP_v3_4_Issue1_2.pdf - 5.2.28), here are logs from smppsim
`2019.09.04 13:38:31 640 INFO    20 short_message=id:0 sub:001 dlvrd:001 submit date:1909041338 done date:1909041338 stat:UNDELIV err:000 Text:test`

## What did you expect to see/happen/not happen?
Expected to somehow get message_state from received deliver_sm in hook, but as i see in logs
`[2019-09-04 16:38:39,461] [9] [INFO] Got smpp_command deliver_sm and log_id , smsc_response CommandStatus(code='ESME_ROK', value=0, description='Success')`
smsc_response is filled with only status, which does not tell much, when we receive deliver_sm

**Comment by komuw** [comment]
Hi @Yolley 

if we changed `BaseHook` so that the `response` method accepted an addition parameter `raw_pdu` that is the full PDU as received from SMSC; would that work for you?
```python
class BaseHook(abc.ABC):
    ...
    @abc.abstractmethod
    async def response(
        self,
        smpp_command: str,
        log_id: str,
        hook_metadata: str,
        smsc_response: "state.CommandStatus",
        raw_pdu: bytes,
    ) -> None:
        """
        called after a response is received from SMSC.

        Parameters:
            smpp_command: any one of the SMSC commands eg submit_sm_resp
            log_id: an ID that a user's application had previously supplied to naz to track/correlate different messages.
            hook_metadata: a string that a user's application had previously supplied to naz that it may want to be correlated with the log_id.
            smsc_response: the response from SMSC.
            raw_pdu: he full PDU as received from SMSC
        """
        raise NotImplementedError("response method must be implemented.")
```

So your custom hook would be like:
```python
    async def response(self, smpp_command, log_id, hook_metadata, smsc_response, raw_pdu):
        logger.info(
            f"Got smpp_command {smpp_command} and log_id {log_id}, smsc_response {smsc_response} and raw_pdu {raw_pdu}",
        )
```

and it would log like;
```bash
2019-09-05 17:55:24,976 Got smpp_command enquire_link_resp and log_id zn7vv5s6v1cnqy01v, smsc_response CommandStatus(code='ESME_ROK', value=0, description='Success') and raw_pdu b'\x00\x00\x00\x10\x80\x00\x00\x15\x00\x00\x00\x00\x00\x00\x00\x03'
```

note: the raw_pdu is as is and you would have to parse it yourself to get the items you want

Would that work for you?

**Comment by Yolley** [comment]
> Hi @Yolley
> 
> if we changed `BaseHook` so that the `response` method accepted an addition parameter `raw_pdu` that is the full PDU as received from SMSC; would that work for you?
> 
> ```python
> class BaseHook(abc.ABC):
>     ...
>     @abc.abstractmethod
>     async def response(
>         self,
>         smpp_command: str,
>         log_id: str,
>         hook_metadata: str,
>         smsc_response: "state.CommandStatus",
>         raw_pdu: bytes,
>     ) -> None:
>         """
>         called after a response is received from SMSC.
> 
>         Parameters:
>             smpp_command: any one of the SMSC commands eg submit_sm_resp
>             log_id: an ID that a user's application had previously supplied to naz to track/correlate different messages.
>             hook_metadata: a string that a user's application had previously supplied to naz that it may want to be correlated with the log_id.
>             smsc_response: the response from SMSC.
>             raw_pdu: the full PDU as received from SMSC
>         """
>         raise NotImplementedError("response method must be implemented.")
> ```
> 
> So your custom hook would be like:
> 
> ```python
>     async def response(self, smpp_command, log_id, hook_metadata, smsc_response, raw_pdu):
>         logger.info(
>             f"Got smpp_command {smpp_command} and log_id {log_id}, smsc_response {smsc_response} and raw_pdu {raw_pdu}",
>         )
> ```
> 
> and it would log like;
> 
> ```shell
> 2019-09-05 17:55:24,976 Got smpp_command enquire_link_resp and log_id zn7vv5s6v1cnqy01v, smsc_response CommandStatus(code='ESME_ROK', value=0, description='Success') and raw_pdu b'\x00\x00\x00\x10\x80\x00\x00\x15\x00\x00\x00\x00\x00\x00\x00\x03'
> ```
> 
> note: the raw_pdu is as is and you would have to parse it yourself to get the items you want
> 
> Would that work for you?

That would be great, thank you!

**Comment by komuw** [comment]
@Yolley 

This is now implemented, try out version `0.6.7` https://pypi.org/project/naz/0.6.7/ 

Note that the api for hooks has changed including the names of the two methods. 
see: https://komuw.github.io/naz/hooks.html
